presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java

```diff
@@ -197,6 +205,51 @@ else if (isEmptyNode(rewrittenRight)) {
             return node.replaceChildren(ImmutableList.of(rewrittenLeft, rewrittenRight));
         }
 
+        @Override
+        public PlanNode visitSequence(SequenceNode node, RewriteContext<Void> context)
+        {
+            List<PlanNode> cteProducers = node.getCteProducers();
+            List<PlanNode> newSequenceChildrenList = new ArrayList<>();
+            // Visit in the order of execution
+            for (int i = cteProducers.size() - 1; i >= 0; i--) {
+                PlanNode rewrittenProducer = context.rewrite(cteProducers.get(i));
+                if (!isEmptyNode(rewrittenProducer)) {
+                    newSequenceChildrenList.add(rewrittenProducer);
+                }
+            }
+            PlanNode rewrittenPrimarySource = context.rewrite(node.getPrimarySource());
+            if (isEmptyNode(rewrittenPrimarySource) || newSequenceChildrenList.isEmpty()) {
+                return rewrittenPrimarySource;
+            }
+            // Reverse order for execution
+            Collections.reverse(newSequenceChildrenList);
+            // Add the primary source at the end of the list
+            newSequenceChildrenList.add(rewrittenPrimarySource);
+            return node.replaceChildren(newSequenceChildrenList);
+        }
+
+        @Override
+        public PlanNode visitCteProducer(CteProducerNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            if (isEmptyNode(rewrittenSource)) {
+                // Remove CTE materialization from session
+                // This will be used to convert consumer to values and in further optimizations
```

Comment:
Where is this further optimization happening?

Reply:
In the same optimizer https://github.com/prestodb/presto/pull/22196/files/687eab33a5c17772abef6710c9cf08b4322a8ad4#diff-2341227ca888731a72c9e28c05277e64930ce13aee3ca6af0d744c1e6923c035